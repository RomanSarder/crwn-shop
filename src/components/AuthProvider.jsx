import React, { useEffect, useState } from 'react'
import { auth, firestore, signOut, makeSignUpWithEmailAndPasswordFunction, makeSignInWithGoogleFunction } from '../firebase/utils'

export var AuthContext = React.createContext({
    currentUser: null
})

export default function AuthProvider({ children }) {
    var [currentUser, setCurrentUser] = useState(null)
    var [currentUserUid, setCurrentUserUid] = useState(null)
    
    useEffect(function subscribeToCurrentAuthenticatedUser () {
        var unsubscribe = auth.onAuthStateChanged(async function getUserProfile (user) {
            if (user) {
                try {
                    var userProfileSnapshot = await firestore
                        .collection('users')
                        .doc(user.uid)
                        .get()
    
                    var userProfileData = userProfileSnapshot.data()
                    if (currentUserUid !== user.uid) {
                        setCurrentUser(userProfileData)
                        setCurrentUserUid(user.uid)
                    }
                } catch (error) {
                    console.log(`Error while trying to get user profile:${error.message}`)
                }
            } else {
                setCurrentUserUid(null)
                setCurrentUser(null)
            }
        })

        return unsubscribe
    })

    var contextValue = {
        currentUser, 
        signOut, 
        makeSignUpFunction: makeSignUpWithEmailAndPasswordFunction,
        signInWithGoogle: makeSignInWithGoogleFunction()
    }

    return (
        <AuthContext.Provider value={contextValue}>
            {children}
        </AuthContext.Provider>
    )
}
