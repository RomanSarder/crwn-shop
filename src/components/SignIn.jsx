import React from 'react'
import styled from 'styled-components'

const StyledSignIn = styled.div``

export default function SignIn() {
    return (
        <StyledSignIn>
            Sign In
        </StyledSignIn>
    )
}
