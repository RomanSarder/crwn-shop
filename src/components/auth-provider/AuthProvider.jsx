import React, { useEffect } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { 
    getAuthInstance, 
    getFirestoreInstance, 
    signOut, 
    makeSignUpWithEmailAndPasswordFunction, 
    signInWithGoogle, 
    signInWithEmailAndPassword 
} from '../../firebase/utils'
import { setUser } from '../../store/user/actions'
import { selectUserId } from '../../store/user/selectors'

export var AuthContext = React.createContext({
    signOut, 
    makeSignUpFunction: makeSignUpWithEmailAndPasswordFunction,
    signInWithGoogle,
    signInWithEmailAndPassword,
})

export function handleAuthStateChanged (dispatch, currentUserId) {
    return async function getUserProfile (user) {
        if (user) {
            if (currentUserId !== user.uid) {
                var firestore = await getFirestoreInstance()
                try {
                    var userProfileSnapshot = await firestore
                        .collection('users')
                        .doc(user.uid)
                        .get()
                    var userProfileData = userProfileSnapshot.data()
                    dispatch(setUser({ ...userProfileData, uid: userProfileSnapshot.id }))
                } catch (error) {
                    console.log(`Error while trying to get user profile:${error.message}`)
                }
            }
        } else {
            dispatch(setUser(null))
        }
    }
}

export default function AuthProvider({ children }) {

    var dispatch = useDispatch()
    var currentUserId = useSelector(selectUserId)
    
    
    useEffect(function subscribeToCurrentAuthenticatedUser () {
        var auth = getAuthInstance()
        var unsubscribe = auth.onAuthStateChanged(handleAuthStateChanged(dispatch, currentUserId))
        return unsubscribe
    }, [currentUserId])

    return (
        <AuthContext.Provider value={{
            signOut, 
            makeSignUpFunction: makeSignUpWithEmailAndPasswordFunction,
            signInWithGoogle,
            signInWithEmailAndPassword,
        }}>
            {children}
        </AuthContext.Provider>
    )
}
