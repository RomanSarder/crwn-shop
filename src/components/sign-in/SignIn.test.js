import { act, render, waitFor } from '@testing-library/react'
import 'jest-styled-components'
import SignIn from './SignIn'
import { AuthContext } from '../auth-provider/AuthProvider'
import userEvent from '@testing-library/user-event'

var providerValueMock = {
    signInWithGoogle: jest.fn(),
    signInWithEmailAndPassword: jest.fn(),
}

function renderSignIn () {
    var renderResult = render(
        <AuthContext.Provider value={providerValueMock}>
            <SignIn />
        </AuthContext.Provider>
    )

    return renderResult
}

it('<SignIn /> should properly render form', () => {
    var { container, getByLabelText, getByRole, getAllByRole, getByText } = renderSignIn()

    expect(container.firstChild).toMatchSnapshot()

    expect(getByRole('heading')).toHaveTextContent('I already have an account')
    expect(getByText((content, element) => {
        return element.tagName.toLowerCase() === 'span' && /^sign in$/i.test(content)
    })).toBeVisible()
    expect(getByRole('textbox')).toHaveAttribute('name', 'email')
    expect(getByLabelText('password', { exact: false })).toBeVisible()
    expect(getByLabelText('password', { exact: false })).toHaveAttribute('type', 'password')

    var buttons = getAllByRole('button')
    expect(buttons.length).toEqual(2)

    expect(buttons[0]).toHaveTextContent(/^sign in$/i)
    expect(buttons[1]).toHaveTextContent(/sign in with google/i)
})

it('<SignIn /> should properly submit form', () => {
    var { getByText } = renderSignIn()

    act(() => {
        var button = getByText((content, element) => {
            return element.tagName.toLowerCase() === 'button' && /^sign in$/i.test(content)
        })
        userEvent.click(button)
    })

    waitFor(() => {
        expect(providerValueMock.signInWithEmailAndPassword).toHaveBeenCalledTimes(1)
    })
})

it('<SignIn /> should properly sign in with google', () => {
    var { getByText } = renderSignIn()

    act(() => {
        var button = getByText((content, element) => {
            return element.tagName.toLowerCase() === 'button' && /^sign in with google$/i.test(content)
        })
        userEvent.click(button)
    })

    waitFor(() => {
        expect(providerValueMock.signInWithGoogle).toHaveBeenCalledTimes(1)
    })
})