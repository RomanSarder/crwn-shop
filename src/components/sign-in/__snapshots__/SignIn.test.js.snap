// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignIn /> should properly render form 1`] = `
.c1 {
  position: relative;
  margin: 4.5rem 0;
}

.c3 {
  color: grey;
  font-size: 1.6rem;
  font-weight: normal;
  position: absolute;
  pointer-events: none;
  left: 0.5rem;
  top: 1rem;
  -webkit-transition: 300ms ease all;
  transition: 300ms ease all;
}

.c4 {
  background: none;
  background-color: white;
  color: grey;
  font-size: 1.8rem;
  padding: 1rem 1rem 1rem 0.5rem;
  display: block;
  width: 100%;
  border: none;
  border-radius: 0;
  border-bottom: 1px solid grey;
  margin: 2.5rem 0;
}

.c4:focus {
  outline: none;
}

.c4:focus ~ .c2 {
  top: -1.4rem;
  font-size: 1.2rem;
  color: black;
}

.c5 {
  min-width: 16.5rem;
  width: auto;
  height: 5rem;
  -webkit-letter-spacing: 0.5px;
  -moz-letter-spacing: 0.5px;
  -ms-letter-spacing: 0.5px;
  letter-spacing: 0.5px;
  line-height: 5rem;
  padding: 0 3.5rem 0 3.5rem;
  font-size: 1.5rem;
  background-color: black;
  color: white;
  text-transform: uppercase;
  font-family: 'Open Sans Condensed';
  font-weight: bolder;
  cursor: pointer;
  border: none;
}

.c5:hover {
  background-color: white;
  color: black;
  border: 1px solid black;
}

.c6 {
  min-width: 16.5rem;
  width: auto;
  height: 5rem;
  -webkit-letter-spacing: 0.5px;
  -moz-letter-spacing: 0.5px;
  -ms-letter-spacing: 0.5px;
  letter-spacing: 0.5px;
  line-height: 5rem;
  padding: 0 3.5rem 0 3.5rem;
  font-size: 1.5rem;
  background-color: black;
  color: white;
  text-transform: uppercase;
  font-family: 'Open Sans Condensed';
  font-weight: bolder;
  cursor: pointer;
  border: none;
  background-color: #4285f4;
  color: white;
}

.c6:hover {
  background-color: white;
  color: black;
  border: 1px solid black;
}

.c6:hover {
  background: #357ae8;
  border: none;
  color: white;
}

.c0 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0 .title {
  font-size: 2.8rem;
  margin: 1rem 0;
}

.c0 span {
  font-size: 2rem;
}

.c0 .buttons {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

@media (min-width:576px) {
  .c0 {
    width: 38rem;
  }
}

<div
  class="c0"
>
  <h2
    class="title"
  >
    I already have an account
  </h2>
  <span>
    Sign In
  </span>
  <form>
    <div
      class="c1"
    >
      <label
        class="c2 c3"
        for="email-1"
      >
        Email
      </label>
      <input
        class="c4"
        id="email-1"
        name="email"
        type="email"
        value=""
      />
    </div>
    <div
      class="c1"
    >
      <label
        class="c2 c3"
        for="password-2"
      >
        Password
      </label>
      <input
        class="c4"
        id="password-2"
        name="password"
        type="password"
        value=""
      />
    </div>
    <div
      class="buttons"
    >
      <button
        class="c5"
        type="submit"
      >
        Sign In
      </button>
      <button
        class="c6"
        type="button"
      >
        Sign In With Google
      </button>
    </div>
  </form>
</div>
`;
